src/test/scala/chiseltest/tests/DecoupledBackPressureInputStallTest.scala

```diff
@@ -0,0 +1,153 @@
+package chiseltest.tests
+
+import chisel3._
+import chisel3.experimental.BundleLiterals.AddBundleLiteralConstructor
+import chisel3.util.{Counter, Decoupled}
+import chiseltest.experimental.TestOptionBuilder.ChiselScalatestOptionBuilder
+import chiseltest.internal.WriteVcdAnnotation
+import chiseltest._
+import org.scalatest.freespec.AnyFreeSpec
+import org.scalatest.matchers.must.Matchers.be
+import org.scalatest.matchers.should.Matchers.convertToAnyShouldWrapper
+
+import scala.util.Random
+
+class GcdInputBundle(val w: Int) extends Bundle {
+  val value1 = UInt(w.W)
+  val value2 = UInt(w.W)
+}
+
+class GcdOutputBundle(val w: Int) extends Bundle {
+  val value1 = UInt(w.W)
+  val value2 = UInt(w.W)
+  val gcd    = UInt(w.W)
+}
+/**
+  * Compute GCD using subtraction method.
+  * Subtracts the smaller of registers x and y from the larger until register y is zero.
+  * value input register x is then the Gcd
+  * returns a packet of information with the two input values and their GCD
+  */
+class DecoupledGcd(width: Int) extends Module {
+
+  val input = IO(Flipped(Decoupled(new GcdInputBundle(width))))
+  val output = IO(Decoupled(new GcdOutputBundle(width)))
+
+  val xInitial    = Reg(UInt())
+  val yInitial    = Reg(UInt())
+  val x           = Reg(UInt())
+  val y           = Reg(UInt())
+  val busy        = RegInit(false.B)
+
+  input.ready := ! busy
+  output.bits := DontCare
+  output.valid := false.B
+
+  when(busy)  {
+    // during computation keep subtracting the smaller from the larger
+    when(y > 0.U) {
+      when(x > y) {
+        x := x - y
+      }.otherwise {
+        y := y - x
+      }
+    }.otherwise {
+      // when y becomes zero computation is over, signal valid data to output
+      output.bits.value1 := xInitial
+      output.bits.value2 := yInitial
+      output.bits.gcd := x
+      output.valid := true.B
+      busy := ! output.ready
+    }
+  }.otherwise {
+    when(input.valid) {
+      // valid data available and no computation in progress, grab new values and start
+      val bundle = input.deq()
+      x := bundle.value1
+      y := bundle.value2
+      xInitial := bundle.value1
+      yInitial := bundle.value2
+      busy := true.B
+      output.valid := false.B
+    }
+  }
+
+  val cycles = IO(Output(UInt(32.W)))
```

Comment:
In a proper design, this wouldn't be part of the DUT right? This is pure test instrumentation? Can this be moved into the test harness, or if a cycle count isn't available yet (is it?) can we have a TODO for it?